module/zfs/zio.c

```diff
@@ -3245,16 +3245,36 @@ zio_ddt_write(zio_t *zio)
 	ddt_t *ddt = ddt_select(spa, bp);
 	ddt_entry_t *dde;
 	ddt_phys_t *ddp;
+	boolean_t added = B_FALSE;
 
 	ASSERT(BP_GET_DEDUP(bp));
 	ASSERT(BP_GET_CHECKSUM(bp) == zp->zp_checksum);
 	ASSERT(BP_IS_HOLE(bp) || zio->io_bp_override);
 	ASSERT(!(zio->io_bp_override && (zio->io_flags & ZIO_FLAG_RAW)));
 
 	ddt_enter(ddt);
-	dde = ddt_lookup(ddt, bp, B_TRUE);
+	if (spa->spa_dedup_max_entries &&
+	    spa->spa_dedup_entries >= spa->spa_dedup_max_entries) {
+		/* Over the desired DDT size, set 'nogrow' flag */
+		dde = ddt_lookup(ddt, bp, B_TRUE, NULL);
+	} else {
+		dde = ddt_lookup(ddt, bp, B_FALSE, &added);
+	}
+	if (dde == NULL) {
+		zp->zp_dedup = B_FALSE;
+		BP_SET_DEDUP(bp, B_FALSE);
+		ASSERT(!BP_GET_DEDUP(bp));
+		zio->io_pipeline = ZIO_WRITE_PIPELINE;
```

Comment:
don't you need to reset the `zio->io_stage` to `ZIO_STAGE_OPEN` here?

Reply:
I don't believe so; this is the same failure path taken elsewhere.

Reply:
I have a vague memory about the need to set `io_stage` to open and/or `zio->io_bp_override` back to `NULL` back when I looked at this years ago but I'll admit I don't remember any of the reasoning behind that.